### ISSUE TITLE
Should we remove the go_repository rule?

### ISSUE THREAD

**Issue by ianthehat** [bazelbuild/rules_go#460]
*TLDR:* go_repository is overly complex, and causes problems for efficient distributed build systems, and we shouldn't really need it.

# Why does it exist?
The go repository rule does two main things that could _not_ be encompassed by a non go specific rule.
* Understands go package import redirection
* Generates BUILD files by running gazelle after download

# Why is it a problem?
Running arbitrary programs at workspace fetch time is not very friendly to systems using `bazel fetch` to prefill an image before running on isolated servers.

Running programs that use proprietary protocols to fetch content from the internet directly is not good for companies that want cached and verified copies of external dependencies, and run their entire build systems with only access to that cache.

Attempting to build those programs from source during that process is both very awkward (you cannot use bazel to build them for instance) and also not friendly to systems that don't expect `blaze fetch` to not do build work. There is also no way to not build the tools if they are not used, as you are not part of a build graph. There are lots of things it would be nice for gazelle to do in the future that will increase its dependencies considerably, and managing these while still building in this fashion is going to get harder and harder.

# How would it take to fix it

The goal would be to just use the normal repository rules, new_http_repository, new_git_repository. For this to work there are a few considerations. 

Most of the work of fetch_repo needs to be done at workspace editing time, rather than repository fetch. In other words, given an import path you need to work out the final fully resolved repository url and vcs type **before** you write the workspace rule, so that you can pick the right one.

The normal repository rules do not support adding BUILD files throughout the tree, but only adding a root BUILD file, so we need to generate these full repository BUILD files instead.

The BUILD files must be generated ahead of time, rather than just in time, so they would also be generated as the repository was added to the workspace. This means they have to be checked in somehwere.

# What are any other advantages

The BUILD files for external dependencies are no longer magic and invisible. This means if gazelle cannot fully cope with a dependency, you can hand edit the resulting BUILD file to fix it easily. It also makes it easier for users to examine the rules for their dependencies, and understand the full dependency tree.

It makes it easy to have a repository of BUILD files for common dependencies. While it would always be better to upstream BUILD files into the original repository anywhere that such changes would be accepted, this does make it easier to have a library of BUILD files for repositories where that is not accepted/feasible.

# How would we do it

Until recently doing this was not feasible, but with the current go rules we can now have packages independant of directory layout, and single build files for entire trees of packages.

To support all the same use cases, we would need to add svn and mercurial repository rules to match the git ones, ideally these would not be go rules but standard ones. I would personally like to see a bazelbuild/rules_vcs repository that has skylark git rules in it as well.

Most of the work though is in changing gazelle to both be able to generate these top level only build files, and also allowing it to correctly generate dependencies to the rules in these files. We would also need to change wtool (or better yet merge it into gazelle and then change it) to generate the workspace rules for the go dependencies.

**Comment by pmbethe09** [comment]
I assume you mean both go_repository and new_go_repository?

I will admit I have enjoyed not having to maintain a directory of //third_party/BUILD.someremotedep and so forth, which you have to do for languages other than Go.
If you generate a file, but then hand-edit it, it can be very tricky to 're-generate' and then this checked-in BUILD file gets stale and one day the remote build breaks and issues arise.

Is there no way to rework new_go_repository to support some of the features in new_http_repository, new_git_repository?

**Comment by ianthehat** [comment]
Yes, I would kill them both.

go_repository already has all the features of new_http_repository, it can never get the exact features of new_git_repository because the interface used by java is not exposed to skylark, but I am already proposing to write a new version that sits directly on top of the git binary anyway, so it will have.

The problem is not features of those rules that we don't have, but the fact that we have to build and run both fetch_repo and gazelle from inside a repository_rule, the extra bits we add. If gazelle switches to depend on a more powerful library that understands skylark a lot better, so it can do more complex rewrites, and understand and process go dependancies so it can maintain the workspace file properly, how are we going to fetch and build it in a sane way, it's already a serious pain that relies on the `go build` system, and at the moment it only really depends on buildifier...

Updating hand edited files is exactly what gazelle has to be awesome at, maintaining the build files in your main tree that are full of code generation rules and random other languages is it's primary job.
Also presumably if you hand edit the file, it's because the autogenerated file is insufficient, which means that's a dependancy that gazelle could not handle and thus would never be usable through go_repository anyway...

The only time the build would break is because you are updating the version of the remote repository, otherwise you are locked to a stable version that matches the build file.

If you imagine the day where you say
```
bazel run @rules_go:gazelle sync golang.org/x/image/tiff
```

And it goes and updates the workspace, updating the commit and sha256, rechasing the go redirects in case they have changed, and then updates the _//third_party/BUILD.same_as_the_workspace_rule_name.bazel_ build file for you to match the new content, I don't think there is any extra or avoidable overhead in terms of work for the user.

In my perfect world, you would have a CI system that was doing something equivalent in concept to
```
bazel run @rules_go:gazelle sync --all
bazel run @rules_go:go_fix
bazel test //...
git commit -am "Automatic library version upgrade."
```

to auto migrate all dependancies to head as long as all the tests are still passing!

I agree that there is a cost to have a directory full of BUILD files, but there are also visibility benefits, and standardisation (with approaches that bazel users already understand) benefits that should not be ignored.
I am also hopeful that the plans for repositories of BUILD files for common dependancies take off, at which point we can use gazelle to generate them for lots of repositories, and pick the BUILD files out of that repository instead, which means the only ones actually in your source tree are dependancies that both don't support bazel and are not in common enough use to be in the shared build file repo.

**Comment by pmbethe09** [comment]
I am still skeptical, but you do appear to have thought out many of the issues

**Comment by hochhaus** [comment]
To give a non-Googler perspective (left in 2008), I believe that for bazel / rules_go to gain wide acceptance in the broader community it needs to be as simple as possible to use (especially when pulling in third party code). @ianthehat, some of your proposal (eg: `bazel run @rules_go:gazelle sync golang.org/x/image/tiff`) looks amazing. My concern is that removing `go_repository` and `new_go_repository` prior to your proposal being fully implemented will be a regression that could drive the current user base away. Would it be possible to only remove `go_repository` once your proposed changes have landed?

**Comment by ianthehat** [comment]
To be clear, this is an issue to discuss a design proposal, I have no clear idea when it would get done, and if it is done it will be done in stages. It would be easy to support the go_repository mechanism in parallel with this suggestion, I would just not want to for a long term because of the support burden it would cause.

**Comment by hochhaus** [comment]
@ianthehat Thanks for the clarification.

**Comment by jayconrod** [comment]
I like this idea. It solves a lot of problems for us.

* We would no longer need to build `gazelle` and `fetch_repo` using `go build` inside the repository rules. Currently, we produce an extra copy of these binaries that Bazel is only slightly aware of. This will cause us problems in the future with cross-platform distributed execution.
* It would make external dependencies less magical, since we don't need to run Gazelle at build time. Magic is only good as long as nothing breaks. By using standard repository rules and checking in BUILD files, we avoid a lot of complications.
* It would enable small modifications to BUILD files for dependencies. Gazelle could merge changes, and you could write `# keep` on rules you don't want it to mess with.
* It would make it easier to extend Gazelle to support other languages in the future, since there's less integration with the build.

Of course, this is all predicated on whether we can make this truly convenient. If we can add / update BUILD files and WORKSPACE dependencies with a single Gazelle command, I'm very much in favor of this.

**Comment by pmcalpine** [comment]
> To support all the same use cases, we would need to add svn and mercurial repository rules to match the git ones, ideally these would not be go rules but standard ones. I would personally like to see a bazelbuild/rules_vcs repository that has skylark git rules in it as well.

See: https://github.com/bazelbuild/bazel/issues/3304

**Comment by abergmeier** [comment]
Just to shake things up a little, what about instead having a webservice, which would do the upconversion?
So still having a `go_repository` Macro, which internally just refers to a `http_archive` with an `url` value of `http://bazel.build/gazelle-gen/v10/?archive=https%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel-gazelle%2Freleases%2Fdownload%2F0.10.0%2Fbazel-gazelle-0.10.0.tar.gz`. `http://bazel.build/gazelle-gen/v10` would then of course be the webservice.

Personally I think the reason against a single `BUILD` file is performance, since parallelization of eval is done on the package level. And you totally would nip that in the but, then.

**Comment by jayconrod** [comment]
@abergmeier It sounds like the webservice would be a proxy that does what `go_repository` does now. Or does it do something else? I'm not sure I see the advantage.

I've been meaning to close this issue, since we have some conclusion actually. I mentioned this in the [roadmap](https://github.com/bazelbuild/rules_go/blob/master/roadmap.rst#p2-overlay-repositories), but I should mention it here as well.

We've decided to keep `go_repository`. It's implementation is kind of ugly, but it's a really useful tool, and the maintenance isn't that expensive. Two important changes though:

* The `go_repository` rule will move to `@bazel_gazelle`, out of `@io_bazel_rules_go`. Eventually, we want to avoid rules_go depending on Gazelle, and since Gazelle is the key part of `go_repository`, we need to move it to complete that decoupling. You can already get the moved version with `load("@bazel_gazelle//:def.bzl", "go_repository")`.
* Some repositories will require manual modifications to build files that can't be done with `go_repository`. For these, we'll provide `http_archive` and `git_repository` rules that support an `overlay` attribute, which copies pre-generated build files (or any files, actually) to the repository. You can make modifications and check in the pre-generated build files, just like regular build files. These are also available in `@bazel_gazelle`, but we don't have an easy way to generate or update the pre-generated build files yet.

**Comment by abergmeier** [comment]
The webservice would indeed mostly execute Gazelle. But also it could be just a wrapper around Bazel (+ Gazelle) and thus benefit from caching. Not to mention that latency for using a repo would go down a bit.
Obviously this only works if e.g. Google sponsors the service, since it would probably be very bandwidth intensive.
